### PR TITLE
Subjects

### PR DESCRIPTION
**Pull request by Anthony271828**

### CODE REVIEW

**Comment by maxwai** [comment]
Lock

**Comment by maxwai** [comment]
Unlock
Still TODO:
* when user gets a Study group he needs to get the appropriate Subjects
* allow the user to opt in/out of subjects (only the one that are allowed for the study group)

**Review by maxwai — APPROVED** [review]
@Anthony271828 finished from my side. Look at it when you have time

File: cogs/subjects.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        member: Union[Member, User] = ctx.author
+        roles: list[Role] = member.roles
+
+        subjects: list[Role] = await self.get_possible_subjects(ctx, member, roles)
+
+        if not subjects:
+            return
+
+        if subject not in [subject.name for subject in subjects]:
+            embed = Embed(title="No permission",
```
Comment (Anthony271828) [inline cogs/subjects.py:123]: This should throw an error so that it can be treated as an error.

File: cogs/subjects.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+                          description=f"You can't be assigned to this subject.\n"
+                                      f"Please contact an admin if you think this is a mistake")
+            await ctx.reply(content=f"<@{member.id}>", embed=embed)
+            return
+
+        if subject not in [subject.name for subject in subjects if subject not in roles]:
+            embed = Embed(title="Already assigned",
```
Comment (Anthony271828) [inline cogs/subjects.py:130]: This should throw an error so that it can be treated as an error.

File: cogs/subjects.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        await StudySubjectUtil.update_category_and_separator(role.id, ctx, db, key, msg)
+
+    async def get_possible_subjects(self, ctx: Context,
+                                    member: Union[Member, User],
+                                    roles: list[Role]) -> list[Role]:
+        study_group: list[Role] = [document.role for document in
+                                   await SubjectsOrGroups(self.bot, SubjectsOrGroupsEnum.GROUP).find({}) if
+                                   document.role in roles]
+        if not study_group:
+            embed = Embed(title="No Study group assigned",
```
Comment (Anthony271828) [inline cogs/subjects.py:253]: This should throw an error so that it can be treated as an error.

File: cogs/groupSubjectLinking.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            default: If this link is assigned by default when joining the study group
+        """
+
+        study_role, subject_role = await self.check_mentions(ctx)
+
+        existing_link = [document for document in await self.db.find({}) if
+                         document.group == study_role and document.subject == subject_role]
+
+        if existing_link:
+            existing_link = {
```
Comment (Anthony271828) [inline cogs/groupSubjectLinking.py:89]: This is also provided by the 'document' property.

File: cogs/groupSubjectLinking.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+        study_role, subject_role = await self.check_mentions(ctx)
+
+        link = [document for document in await self.db.find({}) if
+                document.group == study_role and document.subject == subject_role]
+
+        if not link:
+            raise LinkingNotFoundError
+
+        document = {
```
Comment (Anthony271828) [inline cogs/groupSubjectLinking.py:132]: This is also provided by the 'document' property.

File: cogs/subjects.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        member: Union[Member, User] = ctx.author
+        roles: list[Role] = member.roles
+
+        subjects: list[Role] = await self.get_possible_subjects(ctx, member, roles)
+
+        if not subjects:
+            return
+
+        if subject not in [subject.name for subject in subjects if subject in roles]:
+            embed = Embed(title="Not assigned",
```
Comment (Anthony271828) [inline cogs/subjects.py:163]: This should throw an error so that it can be treated as an error.

File: cogs/groupSubjectLinking.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    @bot_chat(bot_channels)
+    @has_guild_permissions(administrator=True)
+    async def link(self, ctx: Context):
+        if not ctx.invoked_subcommand:
+            raise BadArgument
+        member: Union[Member, User] = ctx.author
+        logger.info(f'User="{member.name}#{member.discriminator}({member.id})", Command="{ctx.message.content}"')
+
+    @link.command(pass_context=True)
+    async def add(self, ctx: Context,
```
Comment (Anthony271828) [inline cogs/groupSubjectLinking.py:66]: Procedural question to link roles.
Automatically add links of all higher semesters to make the work easier?

**Review by Anthony271828 — COMMENTED** [review]
Minor changes and one procedural question still to be addressed.

**Pull request merged**